blockchain/fastsync.go

```diff
@@ -212,34 +213,37 @@ func worker(cache *utxoCache, jobs <-chan []byte, results chan<- *result) {
 // downloadUtxoSet will attempt to connect to make an HTTP GET request to the
 // provided sources one at a time and download the UTXO set to the provided path.
 // If a proxy is provided it will use it for the HTTP connection.
-func downloadUtxoSet(sources []string, proxy *socks.Proxy, pth string) error {
-	dialFunc := net.Dial
-	if proxy != nil {
-		dialFunc = proxy.Dial
-	}
-	tr := &http.Transport{
-		Dial: dialFunc,
-	}
-	client := &http.Client{Transport: tr}
+func downloadUtxoSet(sources []string, proxy *socks.Proxy, pth string) (string, error) {
+	var fileName string
+
 	for _, src := range sources {
-		resp, err := client.Get(src)
-		if err != nil {
-			continue
-		}
-		if resp.StatusCode != http.StatusOK {
-			continue
-		}
-		file, err := os.Create(pth)
-		if err != nil {
-			return err
-		}
 		log.Infof("Downloading UTXO set from %s", src)
-		_, err = io.Copy(file, resp.Body)
+		err := retry.Do(
+			func() error {
+				// TODO: add support for proxies to grab.
+				resp, err := grab.Get(pth, src)
+				if err != nil {
+					return err
+				}
+
+				fileName = resp.Filename
+
+				return nil
+			},
+			retry.Attempts(3),
+			retry.RetryIf(func(err error) bool {
+				return !strings.Contains(err.Error(), "connection refused")
+			}),
+		)
+
 		if err != nil {
-			return err
+			log.Errorf("Failed downloading UTXO set: %s", err.Error())
+			continue
 		}
```

Comment:
LGTM -  could you add the src also - that will help if someone is someone suspects a bad on in the set.

Reply:
This isn't needed. The errors include the domain and are VERY clear. It also aggregates them all and shows you all the failures for each retry. =)

Reply:
Got it! Very nice.